### ISSUE TITLE
[🐛 Bug]:  if session initialisation failed, the result of the test is not be part of the reporter

### ISSUE THREAD

**Issue by NIBOSAP** [webdriverio/webdriverio#8307]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md#reporting-new-issues).

### WebdriverIO Version

7.16.11

### Node.js Version

14

### Mode

WDIO Testrunner

### What happened?

We are running more than 100 specs in an execution, with 68 instances, via saucelabs. 
For an unknown reason, some tests at the beginning fail with this error:
```
[0-25] 2022-05-05T12:28:36.873Z ERROR webdriver: MaxRedirectsError: Redirected 10 times. Aborting.
[0-25]     at Request._onResponseBase (/usr/src/app/node_modules/got/dist/source/core/index.js:859:35)
[0-25]     at Request._onResponse (/usr/src/app/node_modules/got/dist/source/core/index.js:940:24)
[0-25]     at ClientRequest.<anonymous> (/usr/src/app/node_modules/got/dist/source/core/index.js:954:23)
[0-25]     at Object.onceWrapper (events.js:520:26)
[0-25]     at ClientRequest.emit (events.js:412:35)
[0-25]     at ClientRequest.emit (domain.js:475:12)
[0-25]     at ClientRequest.origin.emit (/usr/src/app/node_modules/@szmarczak/http-timer/dist/source/index.js:43:20)
[0-25]     at HTTPParser.parserOnIncomingClient (_http_client.js:647:27)
[0-25]     at HTTPParser.parserOnHeadersComplete (_http_common.js:127:17)
[0-25]     at TLSSocket.socketOnData (_http_client.js:515:22)
[0-25] 2022-05-05T12:28:36.893Z ERROR @wdio/runner: Error: Failed to create session.
[0-25] Redirected 10 times. Aborting.
[0-25]     at startWebDriverSession (/usr/src/app/node_modules/webdriver/build/utils.js:72:15)
[0-25]     at processTicksAndRejections (internal/process/task_queues.js:95:5)
[0-25]     at Function.newSession (/usr/src/app/node_modules/webdriver/build/index.js:58:45)
[0-25]     at remote (/usr/src/app/node_modules/webdriverio/build/index.js:73:22)
[0-25]     at Runner._startSession (/usr/src/app/node_modules/@wdio/runner/build/index.js:223:56)
[0-25]     at Runner._initSession (/usr/src/app/node_modules/@wdio/runner/build/index.js:176:25)
[0-25]     at Runner.run (/usr/src/app/node_modules/@wdio/runner/build/index.js:88:19)
[0-25] FAILED in chrome - /e2e/specs/STS/STS_DataRestrictions/TC_DataRestrictionsList.spec.js
```
So, the tests which are getting this error are "failed". But our specific reporter (the same with all reporters like spec, allure) doesn't show the failed tests.

If I understand correctly, it's because the [runnerStat](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-reporter/src/index.ts#L191) is undefined : the **runner:end** is called in the **_shutdown** method [here](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-runner/src/index.ts#L169) and the **runner:start** where we define the **runnerStat** has never been called at this step.

To sum up, if session initialisation failed, the result of the test will not be part of the reporter. 

### What is your expected behavior?

if session initialisation failed, the result of the test should be part of the reporter and should be failed.

### How to reproduce the bug.

You can do: In 'node_modules\webdriver\build\utils.js', replace (in the try/catch) the line 67 by:
```
throw new Error('Failed to create session.\n');
```

### Relevant log output

```typescript
not relevant
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [X] I have searched the existing issues

**Comment by erwinheitzman** [comment]
Hi @NIBOSAP I tried to reproduce this using the information you have shared. However with the spec reporter I do see the correct total amount of passed and failed tests in the report.
Do you mean a different kind of result by any chance?

**Comment by NIBOSAP** [comment]
Hi @erwinheitzman , 
Many thanks to have a look !
The correct total amount cames from the cli: 
https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-cli/src/interface.ts#L298
Not from the spec reporter. I think that above this line in the console you don't see the test in failure.

**Comment by wyhmichael** [comment]
I am experiencing the similar issue as well with wedriverio@7.19.5; in my case, spec reporter works fine, too, but the failure is not in the other reporters, e.g., json reporter, allure reporter

**Comment by christian-bromann** [comment]
> To sum up, if session initialisation failed, the result of the test will not be part of the reporter.

This was by design. For a test to start a remote session is required. If this fails we don't even initialise the framework and therefor no reports are generated. I agree with you though that a reporter should be able to recognise this. I suggest the following fix: make sure that `onRunnerEnd` is called even though a session wasn't getting initialised. I would take a look into the [`wdio-reporter`](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-reporter/src/index.ts#L189-L201) code and make sure that the [`runner:end`](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-runner/src/index.ts#L414-L418) event is called.

@NIBOSAP you want to help us taking a look at this?

**Comment by NIBOSAP** [comment]
@christian-bromann I took a look  and I didn't find a simple way to fix this without lot of refactoring including breaking changes... Anyway, changing this approach will cost a lot of time.

About calling `onRunnerEnd` even though the session wasn't getting initialised, a possible quick fix could be :
https://github.com/webdriverio/webdriverio/pull/8334. Just a draft to show you the idea...
Probably not the best solution :(, and this could cause breaking changes in the others reporters created by the community... but in a very special case, not handled by their reporter also.

**Comment by christian-bromann** [comment]
@NIBOSAP thanks for raising the PR.

If we can guarantee that the change doesn't break existing reporters I am happy to take a look. Also can you provide screenshots how the reporter looks like before and after the change?

**Comment by NIBOSAP** [comment]
@christian-bromann In fact, nothing changes for the spec-reporter, with or without the fix the log will be:
<img width="1163" alt="image" src="https://user-images.githubusercontent.com/54361411/169039149-1dd09702-b773-4557-a587-cd5a6c296d39.png">

The only difference is: in this case of failure, the `onRunnerEnd` will be called in the reporter. 

That means it depends on what each reporter does in the ` onRunnerEnd`. So, **this fix can cause breaking change**.

- `spec-reporter`: nothing change
- `allure-reporter`: nothing change  as the `onRunnerEnd` is not implemented (cf @wyhmichael )
- `dot-reporter`: nothing change as the `onRunnerEnd` is not implemented
- `concise-reporter`: Before the fix, the concise reporter was not called, so nothing showed up. Now, with the fix:

<img width="740" alt="image" src="https://user-images.githubusercontent.com/54361411/169041537-403c2e12-eb71-480f-9c27-72c378976f83.png">  

- `junit-reporter`: Before the fix, the junit reporter was not called, so nothing showed up. Now, with the fix:

<img width="1365" alt="image" src="https://user-images.githubusercontent.com/54361411/169044975-18340d10-158b-46cf-b231-3443d7c1b99f.png">

- `json-reporter` : Before the fix, the json reporter was not called, so nothing showed up. Now with the fix, an error will happen (cf @wyhmichael )

A better way could be to add another property to the `runnerStat` like: `initialisationFailed`. It doesn't solve the potential breaking change, but it helps to handle it properly in the reporters side.

**Comment by christian-bromann** [comment]
@NIBOSAP thanks for the write up!

I think the better solution is then to ensure that the `runner:start` reporter event is called at all times. In case of a successful session initialisation with session id, in case without a session initialisation without session id.

**Comment by NIBOSAP** [comment]
@christian-bromann It's what I tried first: to call the `runner:start` before the [initialise framework](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-runner/src/index.ts#L152)... But it's seems more complicated as the parameter of the `runner:start` needs `browser`, and `browser` is not initialised yet at this step... Because

[here](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-runner/src/index.ts#L191), how the `runner:start` is initialized:
```
this._reporter.emit('runner:start', {
            cid,
            specs,
            config: browser.options,
            isMultiremote,
            instanceOptions: isMultiremote
                ? multiRemoteBrowser.instances.reduce((prev, browserName) => {
                    prev[multiRemoteBrowser[browserName].sessionId] = multiRemoteBrowser[browserName].options;
                    return prev;
                }, {})
                : {
                    [browser.sessionId]: browser.options
                },
            sessionId: browser.sessionId,
            capabilities: isMultiremote
                ? multiRemoteBrowser.instances.reduce((caps, browserName) => {
                    caps[browserName] = multiRemoteBrowser[browserName].capabilities;
                    caps[browserName].sessionId = multiRemoteBrowser[browserName].sessionId;
                    return caps;
                }, {})
                : { ...browser.capabilities, sessionId: browser.sessionId },
            retry: this._specFileRetryAttempts
        });
```
and the failure happens [here](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-runner/src/index.ts#L166).

it's like a snake biting its own tail...

And If we call the `runner:start` without the `browser`, there will be a breaking change too. Maybe more violent. So, I don't know... :(

**Comment by NIBOSAP** [comment]
@christian-bromann , Thanks, with this way: 

- `spec-reporter`: nothing change
- `allure-reporter`: nothing change  as the `onRunnerEnd` is not implemented (cf @wyhmichael )
- `dot-reporter`: nothing change as the `onRunnerEnd` is not implemented
- `concise-reporter`: Before the fix, the concise reporter was not called, so nothing showed up. Now, with the fix:

<img width="740" alt="image" src="https://user-images.githubusercontent.com/54361411/169041537-403c2e12-eb71-480f-9c27-72c378976f83.png">  

- `junit-reporter`: Before the fix, the junit reporter was not called, so nothing showed up. Now, with the fix:

<img width="1061" alt="image" src="https://user-images.githubusercontent.com/54361411/171377416-0031c1e5-06e6-43b1-bd18-e110a85c41bf.png">

- `json-reporter` : Before the fix, the json reporter was not called, so nothing showed up. Now with the fix, it seems better: (cf @wyhmichael )

<img width="1362" alt="image" src="https://user-images.githubusercontent.com/54361411/171378317-ab6d2a79-ad98-44aa-9d94-782098a516b7.png">

I updated the [PR](https://github.com/webdriverio/webdriverio/pull/8334/files) to let you take a look.

**Comment by christian-bromann** [comment]
The was merged and released. Thanks for the contribution!

**Comment by wyhmichael** [comment]
This is available in v7 not v8, right?

**Comment by christian-bromann** [comment]
Yes, it is available in `v7`

**Comment by NIBOSAP** [comment]
@wyhmichael @christian-bromann From my understanding, the fix will be available only in ```v8```, see [commment](https://github.com/webdriverio/webdriverio/pull/8334#issuecomment-1196874229)
